### PR TITLE
#7687 

### PR DESCRIPTION
**Pull request by HeorhiiBoliuk**
Add an opportunity to select a city of Ukraine and paste your API key in block configuration. Errors that server can pass are displayed in logs.

### CODE REVIEW

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 1]
```diff
@@ -11,40 +16,110 @@
...
-      $weather_type = $data['weather'][0]['main'];
-      $weather_data['weather-type'] = $weather_type;
+      try {
+        $url = 'https://api.openweathermap.org/data/2.5/weather?q=' . urlencode($city) . '&appid=' . urlencode($api_key);
+        $response = \Drupal::httpClient()->get($url);
+        $data = json_decode($response->getBody(), TRUE);
+
+      }
+      catch (ClientException $e) {
+        \Drupal::logger('weather_block')->error('Failed to get weather data for city @city: @message', [
```
Comment (KuzyaWkk) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:108]: It is better to use Dependency Injection and create a container for services

**Review by KuzyaWkk — CHANGES_REQUESTED** [review]
Add .css and .svg files in module directory. Use Dependency Injection

**Comment by HeorhiiBoliuk** [comment]
in the last commit deleted a useless rows

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 2]
```diff
@@ -11,40 +19,144 @@
...
+   */
+  protected LoggerChannelFactoryInterface $logger;
+
+  /**
+   * Constructor for WeatherBlock.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, ConfigFactoryInterface $configFactory, LoggerChannelFactoryInterface $logger) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->configFactory = $configFactory;
+    $this->logger = $logger;
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:42]: Better to use constructor promotion https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor.promotion
It is just simpler

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 3]
```diff
@@ -11,40 +19,144 @@
...
+  }
+
+  /**
+   * Container for WeatherBlock.
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:46]: Doesn't really make sense, what container? Block doesn't accept Container with services itself.
You could just use inherited, and parent also describes what this functions is for

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 4]
```diff
@@ -11,40 +19,144 @@
...
+      'Вінниця', 'Дніпро', 'Донецьк', 'Житомир', 'Запоріжжя', 'Івано-Франківськ',
+      'Київ', 'Кропивницький', 'Луганськ', 'Луцьк', 'Львів', 'Миколаїв', 'Одеса',
+      'Полтава', 'Рівне', 'Суми', 'Тернопіль', 'Ужгород', 'Харків', 'Херсон',
+      'Хмельницький', 'Черкаси', 'Чернівці', 'Чернігів', 'Сімферополь',
+    ];
+
+    $form['cities'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Choose a city'),
+      '#options' => array_combine($cities, $cities),
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:95]: not like that is very important, but keys should be machine names, so in English without spaces, etc.
That said these strings should also be in English and wrapped with ->t(), and if you want them in Ukrainian, then translate them using Drupal's translation system.

Translating is not that straightforward, so you could leave it like that.

Also, I don't know if API accepts English names, if not then it is fine.

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 5]
```diff
@@ -11,40 +19,144 @@
...
+      '#default_value' => $this->configuration['cities'] ?? reset($cities),
+      '#description' => $this->t('Select the city.'),
+    ];
+    $form['API - key'] = [
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:99]: use "machine name" naming for form keys.

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 6]
```diff
@@ -11,40 +19,144 @@
...
+    $this->configuration['cities'] = array_map('trim', explode(',', $form_state->getValue('cities')));
+    $this->configuration['key'] = trim($form_state->getValue('API - key'));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function blockValidate($form, FormStateInterface $form_state): void {
+    $api_key = $form_state->getValue('API - key');
+    if (empty($api_key)) {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:122]: #required => TRUE on the textfield element should cover that validation.

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 7]
```diff
@@ -11,40 +19,144 @@
...
+      $form_state->setErrorByName('API - key', $this->t('Api Key can not be empty'));
+      $form_state->setValidationEnforced(TRUE);
+    }
+    else {
+      $this->configuration['key'] = $api_key;
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:127]: changing the configuration in the validate function doesn't sound good.

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 8]
```diff
@@ -11,40 +19,144 @@
...
 
     foreach ($cities as $city) {
-      $url = 'https://api.openweathermap.org/data/2.5/weather?q=' . $city . '&appid=' . $api_key;
-      $json = file_get_contents($url);
-      $data = json_decode($json, TRUE);
-      $weather_type = $data['weather'][0]['main'];
-      $weather_data['weather-type'] = $weather_type;
+      try {
+        $url = 'https://api.openweathermap.org/data/2.5/weather?q=' . urlencode($city) . '&appid=' . urlencode($api_key);
+        $response = \Drupal::httpClient()->get($url);
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:140]: Use dependency injection for httpClient

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 9]
```diff
@@ -57,15 +169,29 @@ private function getDataFromAPI(): array {
...
       }
     }
     return $formatted_data;
   }
 
+  /**
+   * Get default cities.
+   */
+  private function getDefaultCities(): array {
+    return $this->configFactory->get('weather_block.settings')->get('cities');
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:187]: I don't see wether_block.settings config, and is this function even used?

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 10]
```diff
@@ -57,15 +169,29 @@ private function getDataFromAPI(): array {
...
+  }
+
+  /**
+   * Get default api key.
+   */
+  private function getDefaultApiKey(): array {
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:193]: Same as getDefaultCities

File: web/modules/custom/weather_block/weather_block.info.yml [hunk 11]
```diff
@@ -3,3 +3,6 @@ type: module
 description: 'Displays weather for selected city'
 core_version_requirement: ^9 || ^10
 package: Custom
+
+libraries:
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/weather_block.info.yml:7]: I'm not even sure modules do anything with `libraries` configuration in info.yml

File: web/modules/custom/weather_block/weather_block.module [hunk 12]
```diff
@@ -24,3 +24,10 @@ function weather_block_theme(): array {
   ];
 }
 
+/**
+ * Implements hook_page_attachments_alter().
+ */
+function weather_block_page_attachments_alter(array &$attachments) {
+  $attachments['#attached']['library'][] = 'weather_block/weather_block-css';
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/weather_block.module:31]: You can do that in build function of your block

File: web/modules/custom/weather_block/weather_block.schema.yml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+block.settings.hello_block:
+  type: block_settings
+  label: 'Weather block settings'
+  mapping:
+    hello_block_name:
+      type: string
+      label: 'Weather Block'
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/weather_block.schema.yml:7]: If you are not willing to provide the correct schema of configuration, then do not create it (the schema.yml) at all.
It is not required

**Review by lavrenchuk-dev — APPROVED** [review]

**Review by lavrenchuk-dev — CHANGES_REQUESTED** [review]
Approved my mistake, see comments above for requested changes

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 14]
```diff
@@ -111,17 +108,16 @@ public function blockForm($form, FormStateInterface $form_state) {
...
   /**
    * {@inheritdoc}
    */
   public function blockValidate($form, FormStateInterface $form_state): void {
-    $api_key = $form_state->getValue('API - key');
+    $api_key = $form_state->getValue(['settings', 'key']);
     if (empty($api_key)) {
-      $form_state->setErrorByName('API - key', $this->t('Api Key can not be empty'));
-      $form_state->setValidationEnforced(TRUE);
+      $form_state->setErrorByName('settings][key', $this->t('Api Key can not be empty'));
```
Comment (KuzyaWkk) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:120]: I don't know if it works that way

**Review by KuzyaWkk — APPROVED** [review]

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 15]
```diff
@@ -71,11 +59,15 @@ public function build(): array {
...
     $formatted_data = $this->formatData($api_data);
 
-    return [
+    $build = [
       '#theme' => 'weather_block_template',
       '#content' => $formatted_data,
       '#weather_type' => $weatherType,
     ];
+
+    $build['#attached']['library'][] = 'weather_block/weather_block-css';
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:68]: FYI, nothing prevents you to leave 
```
return [
...
'#attached' => ...
]
```

File: web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php [hunk 16]
```diff
@@ -96,13 +88,18 @@ public function blockForm($form, FormStateInterface $form_state) {
...
+    $form['settings']['key'] = [
+      '#required' => TRUE,
       '#type' => 'textfield',
       '#title' => $this->t('Your API key'),
       '#default_value' => $this->configuration['key'] ?? 'Your API Key',
       '#description' => $this->t('Enter your API key'),
     ];
 
+    if (!empty($this->configuration['settings']['key'])) {
+      $form['settings']['key']['#default_value'] = $this->configuration['key'];
```
Comment (lavrenchuk-dev) [inline web/modules/custom/weather_block/src/Plugin/Block/WeatherBlock.php:100]: `'#default_value' => $this->configuration['key'] ?? NULL`
directly in the array above

**Review by lavrenchuk-dev — CHANGES_REQUESTED** [review]
A 2 little notes, feel free to merge and close the task after fix

**Pull request merged**